packages/basic-crawler/src/internals/error-tracker.ts

```diff
@@ -0,0 +1,310 @@
+} = {}) => {
+    const result: Record<string, unknown> = Object.create(null);
+
+    const extractPathFromStackTraceLine = (line: string) => {
+        const lastStartingRoundBracketIndex = line.lastIndexOf('(');
+
+        if (lastStartingRoundBracketIndex !== -1) {
+            const closingRoundBracketIndex = line.indexOf(')', lastStartingRoundBracketIndex);
+
+            if (closingRoundBracketIndex !== -1) {
+                return line.slice(lastStartingRoundBracketIndex + 1, closingRoundBracketIndex);
+            }
+        }
+
+        return line;
+    };
+
+    // https://v8.dev/docs/stack-trace-api#appendix%3A-stack-trace-format
+    const getPathFromStackTrace = (stack: string[]) => {
+        for (const line of stack) {
+            const path = extractPathFromStackTraceLine(line);
+
+            if (
+                path.startsWith('node:')
+                || path.includes('/node_modules/')
+                || path.includes('\\node_modules\\')
+            ) {
+                continue;
+            }
+
+            return path;
+        }
+
+        return extractPathFromStackTraceLine(stack[0]);
+    };
+
+    const getStackTraceGroup = (error: ErrnoException, storage: Record<string, unknown>) => {
+        const stack = error.stack?.split('\n').map((line) => line.trim());
+
+        let sliceAt = -1;
+
+        if (stack) {
+            for (let i = 0; i < stack.length; i++) {
+                if (stack[i].startsWith('at ') || stack[i].startsWith('eval at ')) {
+                    sliceAt = i;
+                    break;
+                }
+            }
+        }
+
+        let normalizedStackTrace = null;
+        if (sliceAt !== -1) {
+            normalizedStackTrace = showFullStack ? stack!.slice(sliceAt).map((x) => x.trim()).join('\n') : getPathFromStackTrace(stack!.slice(sliceAt));
+        }
+
+        if (!normalizedStackTrace) {
+            normalizedStackTrace = 'null';
+        }
+
+        if (!(normalizedStackTrace in storage)) {
+            storage[normalizedStackTrace] = Object.create(null);
+        }
+
+        return storage[normalizedStackTrace] as Record<string, unknown>;
+    };
+
+    const getErrorCodeGroup = (error: ErrnoException, storage: Record<string, unknown>) => {
+        let { code } = error;
+
+        if (code === undefined) {
+            code = 'null';
+        }
+
+        if (!(code in storage)) {
+            storage[code] = Object.create(null);
+        }
+
+        return storage[code] as Record<string, unknown>;
+    };
+
+    const getErrorNameGroup = (error: ErrnoException, storage: Record<string, unknown>) => {
+        const { name } = error;
+
+        if (!(name in storage)) {
+            storage[name] = {};
+        }
+
+        return storage[name] as Record<string, unknown>;
+    };
+
+    // A is the current one, B may already contain placeholders
+    // This is just a precondition
+    const isSimilarMessage = (a: string, b: string) => {
```

Comment:
Actually we can remove this. This just slows the algorithm down in most cases. The distance precondition would not work in cases like:

```
Invalid URL: https://example.com/some/path/to/a/file.html
Invalid URL: https://anotherone.com/different/one/to/anything.mp4
```

If we rejected merging if `(difference / length) < 0.5` then it wouldn't generalize errors like this because the URL is much longer than the actual message. A better solution would be to check the amount of placeholders after calculating it.

For example, the amount of placeholders shouldn't be greater than the amount of common words (with order in mind).